### ISSUE TITLE
shared X socket vs. X over tcp -> GPU acceleration doesn't work with X over tcp

### ISSUE THREAD

**Issue by mviereck** [docker/docker#27497]
I'm working on a project to provide secure sandboxing for dockered GUI applictions: 
https://github.com/mviereck/x11docker
The core idea is to run a second X server for use with docker. This avoids breaking container isolation like it is the case in widespread solutions with shared X socket from display :0.

GUI clients in docker get access to the second X server over tcp instead of sharing X socket. I often see rendering issues if using shared X socket, that doesn't occure if connecting to X over tcp.
This is especially the case if I share a persistent home folder with docker images, but can occure without this, too. I'm curious about the reasons and prefer X over tcp to avoid this problem.

Now I am working on providing hardware acceleration to docker applications. This works fine using shared X socket, but does not work with X over tcp. 
At first I thought of a tcp protocol problem, but this seems not be the case, because host applications connecting to the second X server over tcp can use hardware acceleration.

To show you the differences, I created a second X server with a dockered terminal and a host terminal in it, checking the OpenGL renderer string in four cases:

```
root@7c7cf556453d:/# # docker connected to X over tcp
root@7c7cf556453d:/# DISPLAY=172.17.42.1:1.0 glxinfo | grep "OpenGL renderer"
OpenGL renderer string: Gallium 0.4 on llvmpipe (LLVM 3.8, 128 bits)
root@7c7cf556453d:/# 
root@7c7cf556453d:/# # docker connected to X with shared X socket
root@7c7cf556453d:/# DISPLAY=:1.0 glxinfo | grep "OpenGL renderer"
OpenGL renderer string: Gallium 0.4 on AMD MULLINS (DRM 2.39.0, LLVM 3.8.0)
root@7c7cf556453d:/# 


lauscher@localhost:~$ # host application on same X server as docker
lauscher@localhost:~$ # host application connected to X over tcp
lauscher@localhost:~$ DISPLAY=172.17.42.1:1.0 glxinfo | grep "OpenGL renderer"
OpenGL renderer string: Gallium 0.4 on AMD MULLINS
lauscher@localhost:~$ 
lauscher@localhost:~$ # host application connected to X using X socket
lauscher@localhost:~$ DISPLAY=:1.0 glxinfo | grep "OpenGL renderer"
OpenGL renderer string: Gallium 0.4 on AMD MULLINS
lauscher@localhost:~$ 
```

As you can see, host applications can get hardware acceleration connected over tcp, dockered applications can not. Accessing X over X socket allows dockered and host applications to use hardware acceleration. 

For testing purpuoses I can run docker --privileged and can set xhost + on the second X server to make sure it is not a permission issue, but it does not help.

Do you have any clue  what may be missing? As hardware acceleration in docker is working with shared X socket, it cannot be due to missing drivers or device files. (I'm sharing all device files in /dev/dri with option --device).

To reproduce the test shown above, you can do this:
- get x11docker
- run 'dpkg-reconfigure x11-common' and choose option 'anybody' to allow yourself to run a second X server
- install openbox, or use another window manager with context menu (f.e. fluxbox)
- use 'x11docker-gui -d' to see developer options, or directly use this command:

`x11docker --X11 --sharenewxsocket --sharegpu --privileged --xhost --wm "openbox" --verbose --  x11docker/xfce  xfce4-terminal`
- You get xfce4-terminal from within docker image. Use context menu of window manager to start a host terminal, too.

To see rendering issues using shared X socket, you can do the same, but additionally share a home folder in ~/.local/share/x11docker with command:
`x11docker --X11  --sudouser --home --xsocket --sharegpu --desktop --verbose --  x11docker/xfce  startxfce4`
This command runs xfce desktop in docker. Start it more than one time to get rendering issues.

I believe it may be a small missing link only to get hardware acceleration with X over tcp in docker working. Do you have an idea?

Output of docker version: 

```
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.3.3
Git commit (client): 7c8fca2
OS/Arch (client): linux/amd64
Server version: 1.6.2
Server API version: 1.18
Go version (server): go1.3.3
Git commit (server): 7c8fca2
OS/Arch (server): linux/amd64

```

Output of docker info:

```
Containers: 43
Images: 423
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 509
 Dirperm1 Supported: true
Execution Driver: native-0.2
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
CPUs: 4
Total Memory: 6.714 GiB
Name: localhost
ID: MBK6:T5AD:ON4F:SKRF:KKGQ:ODYM:WT4Y:DKH2:5WWG:C523:EZT4:DVLZ
WARNING: No memory limit support
WARNING: No swap limit support
```

**Comment by justincormack** [comment]
I think DRI may require passing file descriptors for shared buffers over the X11 socket; you can't do that with TCP.

**Comment by mviereck** [comment]
> I think DRI may require passing file descriptors for shared buffers over the X11 socket; you can't do that with TCP.

That may be a possibility, also I'm not deep enough into X mystery to know about this. But if that would be the case, host applications accessing X over tcp should not be able to access GPU, too. 

But GPU acceleration over tcp works for host applications, as you can see in the terminal output above. I'm already sharing the X socket to docker although I connect over tcp, so the socket would be accessable like it is to host applications. 

I'm quite curious because I cannot see any setup differences for host and for docker applications connected over tcp.

**Comment by mviereck** [comment]
I've researched a bit more and found out:
- GPU acceleration with X over tcp is possible with shared device files in /dev/dri and docker run option --net=host. It is not needed to share X sockets.
- GPU acceleration with shared X sockets is possible with shared device files in /dev/dri and docker run option --ipc=host. Than there are no more rendering issues. I assume it has to do with MIT-SHM extension from X, maybe with memory sharing on GPU memory, too.

In case of shared X sockets, the problem lies within shared memory buffers, as @justincormack assumed for X over tcp. Using X over tcp seems to work in another way and may disable memory sharing, but needs more access to host network than is given from docker by default.

Both options, --net=host and --ipc=host, break container isolation. I have to localize the exact needs for GPU acceleration in both cases and to shrink down these widespread options to what is really needed.

At first place, this to find out is more related to X than to docker, but maybe you have some clues where to look for and which docker options I could check out. Any help and hints are very apprecriated!

**Comment by mviereck** [comment]
I localized the problems with shared X socket. I have to explicitly disable X extension MIT-SHM  to disable usage of shared memory. MIT-SHM seems to be enabled by default in new X servers which I was not aware of. Without MIT-SHM, I have no more rendering glitches and don't need isolation breaking option --ipc=host.
I did not find a solution for GPU acceleration with X over tcp, but don't need it any longer as shared X socket works fine now without any bad RAM access.

This issue can be closed.
